packages/plugin-core/src/features/completionProvider.ts

```diff
@@ -341,6 +341,12 @@ export const provideCompletionItems = sentryReportingCallback(
   }
 );
 
+export const debouncedProvideCompletionItems = _.debounce(
+  provideCompletionItems,
+  100,
+  { trailing: false, leading: true }
```

Comment:
do we actually want it on leading?  This'll potentially make us lose the last keystroke - I think trailing is better in this case? (and 100 ms is probably too fast for a user to perceive lag)

Reply:
You are right, but:

if we remove the leading edge, we lose the trigger focus entirely and any successive keystrokes will not trigger the completion. (imagine mashing [ twice. This triggers the provider twice. if we remove trailing, we lose the completion on the second [ and that's the end.)

we should probably invoke on both edges and play around with the time a bit.

Reply:
> if we remove the leading edge, we lose the trigger focus entirely

I haven't played around much with the VSCode completion provider API, but are you saying we'll hit some VS Code timeout since we didn't provide any data in time and then the provider UI box will no longer show?

Reply:
typing [[asdfasdf]] very quickly, all on 100ms:

1. leading false, trailing true:
 
![image](https://user-images.githubusercontent.com/1219789/174816894-12b9c6d9-e319-4f7a-89da-0f7e0492b411.png)

2. leading true, trailing false:

![image](https://user-images.githubusercontent.com/1219789/174817042-a5f29e3a-adbf-434b-91d9-b3d3ca00651e.png)

3. leading true, trailing true

![image](https://user-images.githubusercontent.com/1219789/174817203-58880966-b9dd-4a02-aec9-23e9d1437937.png)

Reply:
> I haven't played around much with the VSCode completion provider API, but are you saying we'll hit some VS Code timeout since we didn't provide any data in time and then the provider UI box will no longer show?

@jonathanyeung 

I think what's happening is this:

1. we register a few trigger characters, `[` being one of them.
2. the first `[` will trigger the provider
3. the second `[`, if within the debounce window, will not trigger the provider, but will obviously still be typed in the editor.
4. At this point, Intellisense missed the trigger, VSCode saw it as a plain keystroke, didn't find anything in the suggestions, and disengaged. We lose focus here.

Reply:
I see - thanks for the experimentation.  So it looks like doing leading + trailing should get us where we want? 

One other thing we can consider is using the cancellation token: 

```ts
provideCompletionItems(document: TextDocument, position: Position, token: CancellationToken, context: CompletionContext): ProviderResult<T[] | CompletionList<T>>;
```

Reply:
Yes I think for now leading + trailing is sufficient.
I notice that we did at some point use the `resolveCompletionItems` and use cancellation tokens there. Not sure why we stopped using that? I'll follow up with it for the next step.